smac/epm/gaussian_process_mcmc.py

```diff
@@ -0,0 +1,275 @@
+            model.train(X, y, do_optimize=False)
+            self.models.append(model)
+
+        self.is_trained = True
+
+    def loglikelihood(self, theta):
+        """
+        Return the loglikelihood (+ the prior) for a hyperparameter
+        configuration theta.
+
+        Parameters
+        ----------
+        theta : np.ndarray(H)
+            Hyperparameter vector. Note that all hyperparameter are
+            on a log scale.
+
+        Returns
+        ----------
+        float
+            lnlikelihood + prior
+        """
+
+        # Bound the hyperparameter space to keep things sane. Note all
+        # hyperparameters live on a log scale
+        if np.any((-20 > theta) + (theta > 20)):
+            return -np.inf
+            
+        # The last entry is always the noise
+        sigma_2 = np.exp(theta[-1])
+        # Update the kernel and compute the lnlikelihood.
+        self.gp.kernel.vector = np.exp(theta[:-1])
+        
+        try:
+            self.gp.compute(self.X, yerr=np.sqrt(sigma_2))
+        except:
+            return -np.inf
+
+        if self.prior is not None:
+            return self.prior.lnprob(theta) + self.gp.lnlikelihood(self.y, quiet=True)
+        else:
+            return self.gp.lnlikelihood(self.y, quiet=True)
+
+    @BaseModel._check_shapes_predict
+    def predict(self, X_test, **kwargs):
+        r"""
+        Returns the predictive mean and variance of the objective function
+        at X average over all hyperparameter samples.
+        The mean is computed by:
+        :math \mu(x) = \frac{1}{M}\sum_{i=1}^{M}\mu_m(x)
+        And the variance by:
+        :math \sigma^2(x) = (\frac{1}{M}\sum_{i=1}^{M}(\sigma^2_m(x) + \mu_m(x)^2) - \mu^2
+
+        Parameters
+        ----------
+        X_test: np.ndarray (N, D)
+            Input test points
+
+        Returns
+        ----------
+        np.array(N,)
+            predictive mean
+        np.array(N,)
+            predictive variance
+
+        """
+        if not self.is_trained:
+            raise Exception('Model has to be trained first!')
+
+        mu = np.zeros([len(self.models), X_test.shape[0]])
+        var = np.zeros([len(self.models), X_test.shape[0]])
+        for i, model in enumerate(self.models):
+            mu[i], var[i] = model.predict(X_test)
+
+        m = mu.mean(axis=0)
+
+        # See the Algorithm Runtime Prediction paper by Hutter et al.
+        # for the derivation of the total variance
+        v = np.var(mu, axis=0) + np.mean(var, axis=0)
+
+        # Clip negative variances and set them to the smallest
+        # positive float value
+        if v.shape[0] == 1:
+            v = np.clip(v, np.finfo(v.dtype).eps, np.inf)
+        else:
+            v = np.clip(v, np.finfo(v.dtype).eps, np.inf)
+            v[np.where((v < np.finfo(v.dtype).eps) & (v > -np.finfo(v.dtype).eps))] = 0
+
+        return m, v
+    
+    def predict_marginalized_over_instances(self, X_test, **kwargs):
```

Comment:
@KEggensperger @AndreBiedenkapp 
TODOs
1. move to base_gp
2. add docstring
3. add warning in GP facade saying that our GPs don't support instances

Reply:
Done with 6a16ececf5dc1d3101160019c907e8103de9332f